core/src/main/java/io/aiven/klaw/service/TopicSyncControllerService.java

```diff
@@ -1138,4 +1183,79 @@ private boolean checkInPromotionOrder(String envId, String orderOfEnvs) {
     List<String> orderedEnv = Arrays.asList(orderOfEnvs.split(","));
     return orderedEnv.contains(envId);
   }
+
+  private String doesTopicConformToEnvValidation(
+      TopicSyncResponseModel topicRequestReq,
+      String topicPrefix,
+      String topicSuffix,
+      String topicRegex,
+      boolean applyRegex,
+      int maxRepFactor,
+      int maxPartitions) {
+    String validationResponse = "";
+
+    if (!applyRegex) {
+      if (topicPrefix != null
+          && !topicPrefix.isBlank()
+          && !topicRequestReq.getTopicname().startsWith(topicPrefix)) {
+        validationResponse +=
+            String.format(
+                "Topic prefix %s does not match. %s ", topicPrefix, topicRequestReq.getTopicname());
+      }
+
+      if (topicSuffix != null
+          && !topicSuffix.isBlank()
+          && !topicRequestReq.getTopicname().endsWith(topicSuffix)) {
+        validationResponse +=
+            String.format(
+                "Topic suffix %s does not match. %s ", topicSuffix, topicRequestReq.getTopicname());
+      }
+    } else {
+      if (topicRegex != null
+          && !topicRegex.isBlank()
+          && !isRegexAMatch(topicRequestReq, topicRegex)) {
+
+        validationResponse +=
+            String.format(
+                "Topic Regex %s does not match. %s ", topicRegex, topicRequestReq.getTopicname());
+      }
+    }
+
+    if (Integer.parseInt(getValueOrDefault(topicRequestReq.getReplicationfactor(), "0"))
+        > maxRepFactor) {
+      validationResponse +=
+          String.format(
+              "Topic exceeds maximum replication factor %d with %s configured replication factor. ",
+              maxRepFactor, topicRequestReq.getReplicationfactor());
+    }
+    if (topicRequestReq.getTopicpartitions() != null
+        && topicRequestReq.getTopicpartitions().intValue() > maxPartitions) {
+      validationResponse +=
+          String.format(
+              "Topic exceeds maximum partitions %d with %s configured partitions. ",
+              maxPartitions, topicRequestReq.getTopicpartitions().intValue());
+    } else if (topicRequestReq.getTopicpartitions() == null) {
+      validationResponse +=
+          String.format(
+              "Topic partitions not configured. ",
+              maxPartitions,
+              topicRequestReq.getTopicpartitions().intValue());
+    }
+
+    return validationResponse;
+  }
+
+  private static String getValueOrDefault(List<String> params, String defaultValue) {
+    return (params != null && params.size() > 0) ? params.get(0) : defaultValue;
+  }
+
+  private static String getValueOrDefault(String param, String defaultValue) {
+    return (param != null) ? param : defaultValue;
+  }
+
+  private boolean isRegexAMatch(TopicSyncResponseModel topicRequestReq, String topicRegex) {
+    Pattern p = Pattern.compile(topicRegex);
```

Comment:
can this compilation be static ?

Reply:
As the validation can be different for each environment it would mean creating a cache of compiled patterns which I don't know if it would be really worth it.

Reply:
yes will update

Reply:
I didnt make it a static as there could be many options for the regex if there are several environments but, I made it only compile once per list of topics.